### PR TITLE
[2D] Lighting layer batching and code refactoring

### PR DESCRIPTION
**Pull request by UnityPaul**
### Purpose of this PR
- Batch sorting layers affected by the same set of lights.
- Code refactoring.

---
### Testing status

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)

**Automated Tests**: NA
---
### Comments to reviewers
NA

### CODE REVIEW

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs [hunk 1]
```diff
@@ -190,5 +190,41 @@ internal override Shader GetDefaultShader()
...
+
+            const int totalMaterials = 256;
+
+            if(shadowMaterials == null || shadowMaterials.Length == 0)
+                shadowMaterials = new Material[totalMaterials];
+            if(removeSelfShadowMaterials == null || removeSelfShadowMaterials.Length == 0)
+                removeSelfShadowMaterials = new Material[totalMaterials];
+        }
+
+        const int k_NumberOfLightMaterials = 1 << 5 + 3;  // 5 keywords +  volume bit, shape bit
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs:220]: Can we move this line to where `k_NumberOfLightMaterials` is first used?

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs [hunk 2]
```diff
@@ -190,5 +190,41 @@ internal override Shader GetDefaultShader()
...
+
+        // transient data
+        internal Material[] lightMaterials { get; set; }
+        internal Material[] shadowMaterials { get; set; }
+        internal Material[] removeSelfShadowMaterials { get; set; }
+
+        [NonSerialized] internal RenderTargetHandle normalsRenderTarget;
+        [NonSerialized] internal RenderTargetHandle shadowsRenderTarget;
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs:228]: `[NonSerialized]` is not needed here as these are internal.

File: com.unity.render-pipelines.universal/Runtime/2D/Passes/IRenderPass2D.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+using UnityEngine.Rendering;
+using UnityEngine.Rendering.Universal;
```
[the remark this answers is not in the document]
Comment (UnityPaul) [reply com.unity.render-pipelines.universal/Runtime/2D/Passes/IRenderPass2D.cs:3]: fixed

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs (hunk 1 above)
Comment (UnityPaul) [reply com.unity.render-pipelines.universal/Runtime/2D/Renderer2DData.cs:220]: i'll restore it for now, but these things should be use with hashmap

File: com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs [hunk 4]
```diff
@@ -107,55 +130,54 @@ public override void Execute(ScriptableRenderContext context, ref RenderingData
                     combinedDrawSettings.sortingSettings = sortSettings;
                     normalsDrawSettings.sortingSettings = sortSettings;
 
-                    for (int i = 0; i < m_BlendStyleInitialized.Length; ++i)
-                        m_BlendStyleInitialized[i] = false;
-                    
-                    for (int i = 0; i < s_SortingLayers.Length; i++)
+                    var blendStylesCount = m_Renderer2DData.lightBlendStyles.Length;
+                    var hasBeenInitialized = new bool[blendStylesCount];
```
Comment (rainsing) [inline com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs:134]: This needs to be moved out of the Execute() function, or GC alloc test will fail.

Comment (UnityPaul) [reply com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs]: fixed

**Comment by darrenquek** [comment]
Verified smoke test of lost crypt on Editor as well as exploratory testing:

Version: 2020.2.0a19.2490
Revision: trunk 8037ac78137e
Built: Wed, 15 Jul 2020 18:14:55 GMT

**Pull request merged**